### PR TITLE
Add index_fill to ATen XLA tensor

### PR DESCRIPTION
**Pull request by asuhan**

### CODE REVIEW

File: torch_xla/csrc/ops/index_op.cpp [hunk 1]
```diff
@@ -162,6 +163,30 @@ ir::NodePtr IndexPutOp(const ir::Value& buffer, const ir::Value& indices,
       std::move(lower_fn));
 }
 
+ir::NodePtr IndexFillOp(const ir::Value& buffer, xla::int64 dim,
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/index_op.cpp]: ir::Value

File: torch_xla/csrc/xla_lower_util.cpp [hunk 2]
```diff
@@ -288,18 +303,42 @@ xla::XlaOp CreateIndexUpdate(
...
+  xla::XlaComputation combiner_computation =
+      MakeScatterComputation(combiner, buffer_shape.element_type());
   return xla::Scatter(buffer, indices, new_values, combiner_computation,
                       dim_numbers);
 }
 
+xla::XlaOp CreateIndexFill(const xla::XlaOp& buffer, xla::int64 dim,
+                           const xla::XlaOp& index, const xla::XlaOp& value) {
+  xla::Shape buffer_shape = XlaHelpers::ShapeOfXlaOp(buffer);
+
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: line--

File: torch_xla/csrc/xla_lower_util.cpp [hunk 3]
```diff
@@ -288,18 +303,42 @@ xla::XlaOp CreateIndexUpdate(
...
+  xla::ScatterDimensionNumbers dim_numbers;
+  dim_numbers.set_index_vector_dim(1);
+  for (xla::int64 window_dim = 0; window_dim < buffer_shape.rank();
+       ++window_dim) {
+    if (window_dim != dim) {
+      dim_numbers.add_update_window_dims(window_dim);
+    } else {
+      dim_numbers.add_inserted_window_dims(window_dim);
+      dim_numbers.add_scatter_dims_to_operand_dims(dim);
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: I'd use either dim or window_dim in both places here.

File: torch_xla/csrc/xla_lower_util.cpp [hunk 4]
```diff
@@ -288,18 +303,42 @@ xla::XlaOp CreateIndexUpdate(
...
+      MakeScatterComputation(nullptr, buffer_shape.element_type());
+  // Broadcast the value to the right shape required by scatter.
+  xla::Shape index_shape = XlaHelpers::ShapeOfXlaOp(index);
+  std::vector<xla::int64> update_dimensions =
+      xla::util::ToVector<xla::int64>(buffer_shape.dimensions());
+  update_dimensions[dim] = index_shape.dimensions(0);
+  xla::XlaOp updates = xla::Broadcast(
+      xla::ConvertElementType(value, buffer_shape.element_type()),
+      update_dimensions);
+
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: line--

File: torch_xla/csrc/ops/index_op.cpp (hunk 1 above)
Comment (asuhan) [reply torch_xla/csrc/ops/index_op.cpp]: We return `ir::NodePtr` in all helpers.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 2 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 3 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 4 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

File: torch_xla/csrc/xla_lower_util.cpp [hunk 5]
```diff
@@ -72,6 +72,21 @@ std::pair<xla::XlaOp, xla::XlaOp> DotBroadcast(const xla::XlaOp& lhs,
   return std::make_pair(broadcasted_lhs, broadcasted_rhs);
 }
 
+// Builds the computation for the given combiner.
+xla::XlaComputation MakeScatterComputation(
+    const std::function<xla::XlaOp(xla::XlaOp, xla::XlaOp, xla::XlaBuilder*)>&
+        combiner,
+    xla::PrimitiveType element_type) {
+  xla::XlaBuilder cb("scatter-combiner");
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: Can we use "ScatterCombiner"?

Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

**Pull request merged**